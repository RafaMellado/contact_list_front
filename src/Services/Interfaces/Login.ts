export interface LoginRequestBody {
  email: string;
  password: string;
}

export interface LoginResponse {
  token: string;
  username: string;
  exp: string;
}

export interface LoginError {
  error: string;
}
