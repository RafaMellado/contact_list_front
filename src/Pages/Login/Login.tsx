import { FormEvent, useRef } from "react";
import { Button, Card, Form } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import AuthenticationService from "../../Services/AuthenticationService";

export function Login() {
  const { t } = useTranslation();

  const TRANSLATIONS: string = "login";

  const email = useRef<HTMLInputElement>(null);
  const password = useRef<HTMLInputElement>(null);

  const login = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    await AuthenticationService.login({
      email: String(email?.current?.value),
      password: String(password?.current?.value),
    });
  };

  return (
    <div className="d-flex justify-content-center">
      <Card className="login">
        <Card.Body>
          <Form
            onSubmit={login}
            className="d-flex flex-column align-items-center"
          >
            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.email`)}</Form.Label>
              <Form.Control
                as="input"
                type="email"
                placeholder={t<string>(`${TRANSLATIONS}.email`)}
                ref={email}
              />
            </Form.Group>

            <Form.Group className="mb-3">
              <Form.Label>{t<string>(`${TRANSLATIONS}.password`)}</Form.Label>
              <Form.Control
                as="input"
                type="password"
                placeholder={t<string>(`${TRANSLATIONS}.password`)}
                ref={password}
              />
            </Form.Group>

            <Button variant="primary" type="submit">
              {t<string>(`${TRANSLATIONS}.submit`)}
            </Button>
          </Form>
        </Card.Body>
      </Card>
    </div>
  );
}
