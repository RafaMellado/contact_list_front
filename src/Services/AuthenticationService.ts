import Cookies from "universal-cookie";
import { LoginRequestBody, LoginResponse } from "./Interfaces/Login";
import LoginRepository from "./Repository/LoginRepository";

class AuthenticationService {
  async login(data: LoginRequestBody) {
    const response = await LoginRepository.login(data);

    if (response.token) {
      this.setAuthCookie(response);

      window.location.href = "/home";
    }
  }

  setAuthCookie(data: LoginResponse) {
    const cookies = new Cookies();

    cookies.set(
      "user",
      { token: data.token, username: data.username },
      { expires: new Date(data.exp) }
    );
  }
}

export default new AuthenticationService();
