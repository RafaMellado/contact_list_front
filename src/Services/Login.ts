import { LoginData, LoginResponse } from "./Interfaces/Login";
import RestService from "./RestService";
import Cookies from "universal-cookie";

class LoginService {
  async login(loginData: LoginData) {
    const response = await RestService.post("/auth/login", loginData);
    const data: LoginResponse = await response.json();

    if (data.token) {
      this.setCookie(data);
    }

    return !!data.token;
  }

  setCookie(data: LoginResponse) {
    const cookies = new Cookies();

    cookies.set(
      "user",
      { token: data.token, username: data.username },
      { expires: new Date(data.exp) }
    );
  }
}

export default new LoginService();
