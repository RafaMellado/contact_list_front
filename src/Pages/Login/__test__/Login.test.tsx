import { Login } from "../Login";
import AuthenticationService from "../../../Services/AuthenticationService";
import { act, fireEvent, render, waitFor } from "@testing-library/react";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("../../../Services/AuthenticationService");

const mockedAuthenticationService = AuthenticationService as jest.Mocked<
  typeof AuthenticationService
>;

describe(`<${Login.name} />`, () => {
  const factoryComponent = () => render(<Login />);

  test("renders correctly the component", () => {
    const component = factoryComponent();

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("fill correctly and call login", () => {
    const component = factoryComponent();

    act(() => {
      fireEvent.input(component.getByTestId("login-email"), {
        target: {
          value: "test@email.com",
        },
      });

      fireEvent.input(component.getByTestId("login-password"), {
        target: {
          value: "abcdef",
        },
      });

      fireEvent.click(component.getByTestId("login-submit"));

      expect(mockedAuthenticationService.login).toHaveBeenCalledWith({
        email: "test@email.com",
        password: "abcdef",
      });
    });
  });

  test("submit and show errors", async () => {
    mockedAuthenticationService.login.mockRejectedValue({
      error: "unauthorized",
    });

    const component = await factoryComponent();

    fireEvent.click(component.getByTestId("login-submit"));

    await waitFor(() => {
      expect(component.queryByTestId("login-error")).toBeTruthy();
    });

    expect(component.asFragment()).toMatchSnapshot("errors");
  });
});
