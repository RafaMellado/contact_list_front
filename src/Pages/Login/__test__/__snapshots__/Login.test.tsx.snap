// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="login card"
    >
      <div
        class="card-body"
      >
        <form
          class="d-flex flex-column align-items-center"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              login.email
            </label>
            <input
              class="form-control"
              data-testid="login-email"
              placeholder="login.email"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              login.password
            </label>
            <input
              class="form-control"
              data-testid="login-password"
              placeholder="login.password"
              type="password"
            />
          </div>
          <button
            class="btn btn-primary"
            data-testid="login-submit"
            type="submit"
          >
            login.submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Login /> submit and show errors: errors 1`] = `
<DocumentFragment>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="login card"
    >
      <div
        class="card-body"
      >
        <form
          class="d-flex flex-column align-items-center"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              login.email
            </label>
            <input
              class="form-control"
              data-testid="login-email"
              placeholder="login.email"
              type="email"
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              login.password
            </label>
            <input
              class="form-control"
              data-testid="login-password"
              placeholder="login.password"
              type="password"
            />
          </div>
          <div
            class="text-danger my-2"
            data-testid="login-error"
          >
            <span>
              errors.login
            </span>
          </div>
          <button
            class="btn btn-primary"
            data-testid="login-submit"
            type="submit"
          >
            login.submit
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
